IMOD/manpages/checkmtmod.man
.na
.nh
.TH checkmtmod 1 4.6.34 IMOD
.SH NAME
checkmtmod - check various features of a MT model
.SH SYNOPSIS
checkmtmod
.SH DESCRIPTION
Checkmtmod will check a microtubule tracking model built on serial
sections for three kinds of errors: branch points, skipped sections,
and failure of the Z coordinate to increase monotonically. In
addition, it can search for situations where the start of one contour
is near the end of another contour, so that one can then check
whether these contours are in fact the same microtubule.
.P
All branch points are reported.  The Z coordinate is expected to
increase monotonically within each contour, so there will be a report
whenever Z stays the same or decreases.  If an entire contour is in
one Z plane, there will be just one report for that contour.
Z values are expected to occur in a regular progression, and gaps in
that progression will be reported unless they occur at missing
sections, at which no contours have points.
.P
You can specify a list of additional sections on which some points
may be intentionally missing.  The program will then not report the
contours that skip over that section.  You can enter a list with
ranges, such as 1-3,37,78.
.P
To search for starts near ends, enter a distance within the X-Y plane
to search, and the maximum difference in Z coordinates to search.
(Enter / to accept the default values for these parameters, or enter
0,0 to skip the search).
.P
Enter the scaling in Z relative to the X and Y scaling (this would
be the same as the scale value used to display
the model at the true scale).  This value is needed to compute angles
between trajectories of MTs with nearby ends.  If you are not
searching for starts near ends, it doesn't matter what you enter.
.P
When searching for starts near ends, the program will report
the starting and ending points whenever those two points are within
the selected distance in the X-Y plane and the Z value of the
starting point minus the Z value of the ending point is positive and
less than the selected difference in Z values.  The program also
fits a line to the last three points in the contour that ends, and
another line to the first 3 points in the contour that starts.  It
uses these lines to extrapolate the trajectory of each contour to the
position in Z midway between the ending and starting points.  It then
reports the distance between these extrapolated positions, and the
angle between the trajectories in degrees.
.P
For each match, the program outputs on two lines the following
information:
.P
.nf
# of contour that ends, X, Y, Z coordinates of the ending point,
# of contour that starts, X, Y, Z coordinates of the starting point,
separation between objects extrapolated to a single plane in Z,
and angle between the trajectories of the objects.
.fi
.P
This information on extrapolated separation and angle can be used
to zero in on the matches that need to be examined in the model;
most "matches" can be eliminated without examination because either
the angle or the separation is unreasonably large.
.SH HISTORY
Written by David Mastronarde, 1/21/90, jazzed up 4/32/93, modified for
IMOD 4/25/97
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/checkmtmod.f
*       * * * * CHECKMTMOD * * * * *
c       
c       CHECKMTMOD will check a microtubule tracking model built on serial
c       sections for three kinds of errors: branch points, skipped sections,
c       and failure of the Z coordinate to increase monotonically. In
c       addition, it can search for situations where the start of one contour
c       is near the end of another contour, so that one can then check
c       whether these contours are in fact the same microtubule.
c       
c       All branch points are reported.  The Z coordinate is expected to
c       increase monotonically within each contour, so there will be a report
c       whenever Z stays the same or decreases.  If an entire contour is in
c       one Z plane, there will be just one report for that contour.
c       Z values are expected to occur in a regular progression, and gaps in
c       that progression will be reported unless they occur at missing
c       sections, at which no contours have points.
c       
c       You can specify a list of additional sections on which some points
c       may be intentionally missing.  The program will then not report the
c       contours that skip over that section.  You can enter a list with
c       ranges, such as 1-3,37,78.
c       
c       To search for starts near ends, enter a distance within the X-Y plane
c       to search, and the maximum difference in Z coordinates to search.
c       (Enter / to accept the default values for these parameters, or enter
c       0,0 to skip the search).
c       
c       Enter the scaling in Z relative to the X and Y scaling (this would
c       be the same as the scale value used to display
c       the model at the true scale).  This value is needed to compute angles
c       between trajectories of MTs with nearby ends.  If you are not
c       searching for starts near ends, it doesn't matter what you enter.
c       
c       When searching for starts near ends, the program will report
c       the starting and ending points whenever those two points are within
c       the selected distance in the X-Y plane and the Z value of the
c       starting point minus the Z value of the ending point is positive and
c       less than the selected difference in Z values.  The program also
c       fits a line to the last three points in the contour that ends, and
c       another line to the first 3 points in the contour that starts.  It
c       uses these lines to extrapolate the trajectory of each contour to the
c       position in Z midway between the ending and starting points.  It then
c       reports the distance between these extrapolated positions, and the
c       angle between the trajectories in degrees.
c       
c       For each match, the program outputs on two lines the following
c       information:
c       
c       # of contour that ends, X, Y, Z coordinates of the ending point,
c       # of contour that starts, X, Y, Z coordinates of the starting point,
c       separation between objects extrapolated to a single plane in Z,
c       and angle between the trajectories of the objects.
c       
c       This information on extrapolated separation and angle can be used
c       to zero in on the matches that need to be examined in the model;
c       most "matches" can be eliminated without examination because either
c       the angle or the separation is unreasonably large.
c       
c       David Mastronarde  1/21/90, jazzed up 4/32/93
c       modified for IMOD 4/25/97
c       
      use fortmodel

      logical exist,readw_or_imod,zused(10000)
      character*80 modelfile
      integer*4 izskipok(200)
      real*4 xfit(10),yfit(10),zfit(10)
      data distmax/100./,zsepmax/5./
      call setExitPrefix('ERROR: CHECKMTMOD - ')
c       
c       get model
c       
10    write(*,'(1x,a,$)')'Name of model file: '
      read(*,'(a)')modelfile
15    exist=readw_or_imod(modelfile)
      if(.not.exist)then
        print *,'no good, try again'
        go to 10
      endif
c       
      print *,'Skipped points will not be reported for sections that',
     &    ' have no points at all.'
      print *,'Enter list of additional section numbers for which',
     &    ' skipped points',' should not be',
     &    '     reported (ranges OK, Return for none)'
      call rdlist(5,izskipok,nzskipok)
c       
      write(*,'(1x,a,/,a,/,a,f4.0,a,f3.0,a,$)')'Maximum distance in'//
     &    ' X-Y plane and maximum separation in Z to search'//
     &    ' for','   starting points near ending points, or 0,0 for'
     &    //' no search','    [/ for ',distmax,',',zsepmax,']: '
      read(*,*)distmax,zsepmax
c       
      write(*,'(1x,a,$)')'Z scaling relative to X and Y (needed to'//
     &    ' compute angles): '
      read(5,*)zscal
      distsqr=distmax**2
c       
c       find min and max z
c       
      izmin=100000
      izmax=-100000
      do iobj=1,max_mod_obj
        ibase=ibase_obj(iobj)
        do i=1,npt_in_obj(iobj)
          jpnt=abs(object(i+ibase))
          iz=nint(p_coord(3,jpnt))
          izmin=min(izmin,iz)
          izmax=max(izmax,iz)
        enddo
      enddo
c       
c       find which z values are used
c       
      nz=izmax+1-izmin
      do i=1,nz
        zused(i)=.false.
      enddo
c       
      do iobj=1,max_mod_obj
        ibase=ibase_obj(iobj)
        do i=1,npt_in_obj(iobj)
          jpnt=abs(object(i+ibase))
          iz=nint(p_coord(3,jpnt))
          zused(iz+1-izmin)=.true.
        enddo
      enddo
c       
c       take out the ones that are ok to skip
c       
      do iok=1,nzskipok
        ind=izskipok(iok)+1-izmin
        if(ind.gt.0.and.ind.le.nz)zused(ind)=.false.
      enddo
c       
c       now go through objects looking for problems
c       
      write(*,110)
110   format(/,' O   2, C   20 (   65)  means IMOD object 2, ',
     &    'contour 20, WIMP object 65',/)
      do iobj=1,max_mod_obj
        call objtocont(iobj,obj_color,imodobj,imodcont)
        ibase=ibase_obj(iobj)
        ifanyz=0
        do ipt=1,npt_in_obj(iobj)
          jpnt=abs(object(ipt+ibase))
          iz=nint(p_coord(3,jpnt))
          if(ipt.gt.1)then
            if(iz.ne.izlast)ifanyz=ifanyz+1
          endif
          izlast=iz
        enddo
        if(ifanyz.gt.0.or.npt_in_obj(iobj).eq.0)then
          do ipt=1,npt_in_obj(iobj)
            jpnt=object(ipt+ibase)
            if(jpnt.gt.0)then
              iz=nint(p_coord(3,jpnt))
              if(ipt.gt.1)then
                if(iz.le.izlast)then
                  write(*,101)imodobj,imodcont,iobj,ipt,
     &                (p_coord(j,jpnt),j=1,3)
                else
                  nskip=0
                  do izchk=izlast+1,iz-1
                    if(zused(izchk+1-izmin))nskip=nskip+1
                  enddo
                  if(nskip.gt.0)write(*,102)nskip,imodobj,imodcont,
     &                iobj,ipt, (p_coord(j,jpnt),j=1,3)
                endif
              endif
              izlast=iz
            else
              write(*,103)imodobj,imodcont,iobj,ipt,(p_coord(j,-jpnt),j=1,3)
              izlast=p_coord(3,-jpnt)
            endif
          enddo
        else
          write(*,105)imodobj,imodcont,iobj,npt_in_obj(iobj),izlast
        endif
      enddo
101   format('  Z fails to increase, O',i4,', C',i5,' (',
     &    i5,') pt',i5,', XYZ',2f8.1,f6.0)
102   format(i4,' Z values skipped, O',i4,', C',i5,' (',
     &    i5,') pt',i5,', XYZ',2f8.1,f6.0)
103   format('    Branch point in O',i4,', C',i5,' (',
     &    i5,') pt',i5,', XYZ',2f8.1,f6.0)
105   format('  O',i4,', C',i5,' (',i5,') is planar;',i5,
     &    ' points, Z =',i5)
c       
      write(*,*)
      if(zsepmax.gt.0)then
        print *,'Checking for errors completed.'
        print *,'Checking for object starts near object ends;',
     &      '   Object # and X-Y-Z coordinates are reported next:'
c         
c         go through objects in model, comparing last point of each to
c         first point in all other objects
c         
        do iobj=1,max_mod_obj
          ninobj1=npt_in_obj(iobj)
          call objtocont(iobj,obj_color,imodobj,imodcont)
          if(ninobj1.gt.1)then
            jpnt=abs(object(ninobj1+ibase_obj(iobj)))
            xx=p_coord(1,jpnt)
            yy=p_coord(2,jpnt)
            zz=p_coord(3,jpnt)
            if(zz.ne.p_coord(3,object(1+ibase_obj(iobj))))then
              do job=1,max_mod_obj
                call objtocont(job,obj_color,jmodobj,jmodcont)

                if(npt_in_obj(job).gt.1.and.job.ne.iobj)then
                  ipnt2=abs(object(1+ibase_obj(job)))
                  zstrt=p_coord(3,ipnt2)
                  if(zstrt.ne.p_coord(3,abs(
     &                object(npt_in_obj(job)+ibase_obj(job)))))then
                    if(zstrt.gt.zz.and.
     &                  zstrt-zz.le.zsepmax)then
                      dx=abs(xx-p_coord(1,ipnt2))
                      dy=abs(yy-p_coord(2,ipnt2))
                      if(dx.le.distmax.and.dy.le.distmax.and.
     &                    dx**2+dy**2.le.distsqr)then
c                         
c                         got a match,
c                         
                        zmid=0.5*zscal*(zz+zstrt)
                        nfit=0
                        do ipt=max(1,ninobj1-2),ninobj1
                          jpack=abs(object(ipt+ibase_obj(iobj)))
                          nfit=nfit+1
                          xfit(nfit)=p_coord(1,jpack)
                          yfit(nfit)=p_coord(2,jpack)
                          zfit(nfit)=zscal*p_coord(3,jpack)
                        enddo
                        call lsfit(zfit,xfit,nfit,bx1,ax1,ro)
                        call lsfit(zfit,yfit,nfit,by1,ay1,ro)
                        xtrap1=ax1+bx1*zmid
                        ytrap1=ay1+by1*zmid
c                         
                        nfit=0
                        do ipt=1,min(3,npt_in_obj(job))
                          jpack=abs(object(ipt+ibase_obj(job)))
                          nfit=nfit+1
                          xfit(nfit)=p_coord(1,jpack)
                          yfit(nfit)=p_coord(2,jpack)
                          zfit(nfit)=zscal*p_coord(3,jpack)
                        enddo
                        call lsfit(zfit,xfit,nfit,bx2,ax2,ro)
                        call lsfit(zfit,yfit,nfit,by2,ay2,ro)
                        xtrap2=ax2+bx2*zmid
                        ytrap2=ay2+by2*zmid
                        septrap=sqrt((xtrap1-xtrap2)**2+
     &                      (ytrap1-ytrap2)**2)
                        cosang=(1.+bx1*bx2+by1*by2)/sqrt
     &                      ((1.+bx1**2+by1**2)*(1.+bx2**2+by2**2))
                        ang=acosd(cosang)
                        write(*,104)imodobj,imodcont,iobj,xx,yy,zz,
     &                      jmodobj,jmodcont,job,
     &                      (p_coord(i,ipnt2),i=1,3),septrap,ang
                      endif
                    endif
                  endif
                endif
              enddo
            endif
          endif
        enddo
104     format(/,' O',i4,', C',i5,' (',i5,') ends at ',2f7.0,f6.0,
     &      /,' O',i4,', C',i5,' (',i5,') starts @',2f7.0,f6.0,
     &      ' sep',f5.0,f5.0,'deg')
      endif
c       
      print *, 'CHECKING COMPLETED'
      call exit(0)
      end
